Sources/PackageRegistry/SigningEntityTOFU.swift

```diff
@@ -0,0 +1,264 @@
+        version: Version,
+        signingEntity: SigningEntity?,
+        signerVersions: [SigningEntity: Set<Version>],
+        observabilityScope: ObservabilityScope,
+        completion: @escaping (Result<Bool, Error>) -> Void
+    ) {
+        // Package is never signed.
+        // If signingEntity is nil, it means package remains unsigned, which is OK. (none -> none)
+        // Otherwise, package has gained a signing entity, which is also OK. (none -> some)
+        if signerVersions.isEmpty {
+            return completion(.success(true))
+        }
+
+        // If we get to this point, it means we have seen a signed version of the package.
+
+        // TODO: It's possible that some of the signing entity changes are legitimate:
+        // e.g., change of package ownership, package author decides to stop signing releases, etc.
+        // Instead of failing, we should allow and prompt user to add/replace/remove signing entity.
+
+        // We recorded the version's signer previously
+        if let signerForVersion = signerVersions.signingEntity(of: version) {
+            // The given signer is different
+            // TODO: This could indicate a legitimate change in package ownership
+            guard signerForVersion == signingEntity else {
+                return self.handleSigningEntityChanged(
+                    package: package,
+                    version: version,
+                    latest: signingEntity,
+                    existing: signerForVersion,
+                    observabilityScope: observabilityScope
+                ) { result in
+                    completion(result.tryMap { false })
+                }
+            }
+            // Signer remains the same for the version
+            return completion(.success(false))
+        }
+
+        switch signingEntity {
+        // Is the package changing from one signer to another?
+        case .some(let signingEntity):
+            // Check signer(s) of other version(s)
+            if let otherSigner = signerVersions.keys.filter({ $0 != signingEntity }).first {
+                // There is a different signer
+                // TODO: This could indicate a legitimate change in package ownership
+                self.handleSigningEntityChanged(
+                    package: package,
+                    latest: signingEntity,
+                    existing: otherSigner,
+                    observabilityScope: observabilityScope
+                ) { result in
+                    completion(result.tryMap { false })
+                }
+            } else {
+                // Package doesn't have any other signer besides the given one, which is good.
+                completion(.success(true))
+            }
+        // Or is the package going from having a signer to .none?
+        case .none:
+            let versionSigners = signerVersions.versionSigners
+            // If the given version is semantically newer than any signed version,
```

Comment:
👀 

On one hand I think we are trying to be too smart here, but I also think that it will be pretty common for a package to sign new versions only and leave older versions unsigned.

Reply:
this seems correct and reasonable to assume

Reply:
Yah, I think the switch signed => unsigned is significant enough that it should require manual intervention by the user. Otherwise, any attacker can just publish a new unsigned version to circumvent the entire process.